src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -366,23 +366,32 @@ public int Length
 
                 if (value > m_length)
                 {
-                    // clear high bit values in the last int
-                    int last = GetArrayLength(m_length, BitsPerInt32) - 1;
-                    int bits = m_length % 32;
-                    if (bits > 0)
-                    {
-                        m_array[last] &= (1 << bits) - 1;
-                    }
-
-                    // clear remaining int values
-                    Array.Clear(m_array, last + 1, newints - last - 1);
+                    TrimUpperBits();
                 }
 
                 m_length = value;
-                _version++;
             }
         }
 
+        private void TrimUpperBits()
+        {
+            // clear high bit values in the last int
+            int last = GetArrayLength(m_length, BitsPerInt32) - 1;
+            int bits = m_length % 32;
+            if (bits > 0)
+            {
+                m_array[last] &= (1 << bits) - 1;
+            }
+
+            // clear remaining int values
+            if (last + 1 < m_array.Length)
+            {
+                Array.Clear(m_array, last + 1, m_array.Length - last - 1);
+            }
+
+            _version++;
```

Comment:
Yes, I also thought this seems suspicious, but decided to first go with a straightforward implementation. If this is an issue, I can think of two ways to get the desired behavior without modifying the internal data and incrementing version:

1. Do the trimming after every surface operation that might introduce excessive bits (just like `set_Length`)
2. AFTER the CopyTo copied data into the target array, trim the target array. This would effectively leave the internal storage mechanism as is.

The question now is which one would be preferable.